src/main/scala/viper/gobra/ast/internal/transform/OverflowChecksTransform.scala

```diff
@@ -0,0 +1,146 @@
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) 2011-2020 ETH Zurich.
+
+package viper.gobra.ast.internal.transform
+
+import viper.gobra.ast.internal._
+import viper.gobra.reporting.Source
+import viper.gobra.reporting.Source.Parser.Single
+import viper.gobra.util.TypeBounds.BoundedIntegerKind
+import viper.gobra.util.Violation.violation
+
+/**
+  * Adds overflow checks to programs written in Gobra's internal language
+  */
+object OverflowChecksTransform extends InternalTransform {
+  override def name(): String = "add_integer_overflow_checks"
+
+  override def transform(p: Program): Program = transformMembers(memberTrans)(p)
+
+  private def memberTrans(member: Member): Member = member match {
+    // adds overflow checks per statement that contains subexpressions of bounded integer type and adds assume
+    /// statements at the beginning of a function or method body assuming that the value of an argument (of
+    // bounded integer type) respects the bounds.
+    case f@Function(name, args, results, pres, posts, body) =>
+      Function(name, args, results, pres, posts, body map (computeNewBody(args, _)))(f.info)
+
+    // same as functions
+    case m@Method(receiver, name, args, results, pres, posts, body) =>
+      Method(receiver, name, args, results, pres, posts, body map (computeNewBody(args, _)))(m.info)
+
+    // Adds pre-conditions stating the bounds of each argument and a post-condition to check if the body expression
+    // overflows
+    case f@PureFunction(name, args, results, pres, posts, body) => body match {
+      case Some(expr) =>
+        val newPre = pres ++ getPureBlockPres(args)
+        val newPost = posts ++ getPureBlockPosts(expr, results)
+        PureFunction(name, args, results, newPre, newPost, body)(f.info)
+      case None => f
+    }
+
+    // Same as pure functions
+    case m@PureMethod(receiver, name, args, results, pres, posts, body) => body match {
+      case Some(expr) =>
+        val newPre = pres ++ getPureBlockPres(args)
+        val newPost = posts ++ getPureBlockPosts(expr, results)
+        PureMethod(receiver, name, args, results, newPre, newPost, body)(m.info)
+      case None => m
+    }
+
+    /* As discussed on the Gobra meeting (27/10/2020), overflow checks should not be added to predicates, assertions
+    * and any other purely logical (i.e. non-executable code) statements and expressions. This seems to be the approach taken
+    * by other verification tools such as FramaC, as noted by Wytse
+    */
+
+    case x => x
+  }
+
+  /**
+    * Adds overflow checks to the bodies of a block. If block is parametrized,
+    * `computeNewBody` adds assumptions about the parameters' bounds
+    */
+  private def computeNewBody(args: Vector[Parameter.In], body: Block): Block = {
+    val assumeArgBounds = args map { case p@Parameter.In(_, typ) => Assume(assertionExprInBounds(p, typ)(p.info))(p.info) }
```

Comment:
I mentioned this a bunch of times, but think that just adding the information that the parameter are within their bounds is too weak too show many properties. But again, we can change this later, when we see that stronger assumptions are necessary. It is an ok first step.